src/lune/builtins/datetime/mod.rs

```diff
@@ -0,0 +1,222 @@
+                    // HACK: 10 * (timestamp.fract().to_string().len() - 2) gives us the number of digits
+                    // after the decimal
+                }
+            };
+
+            Ok(DateTime::from_unix_timestamp(timestamp_kind, timestamp))
+        })?
+        .with_function("fromUniversalTime", |lua, date_time: LuaValue| {
+            Ok(DateTime::from_universal_time(DateTimeBuilder::from_lua(date_time, lua).ok()))
+        })?
+        .with_function("toUniversalTime", |_, this: DateTime| {
+            Ok(this.to_universal_time())
+        })?
+        .with_function("fromLocalTime", |lua, date_time: LuaValue| {
+            Ok(DateTime::from_local_time(DateTimeBuilder::from_lua(date_time, lua).ok()))
+        })?
+        .with_function("toLocalTime", |_, this: DateTime| {
+            Ok(DateTime::to_local_time(&this))
+        })?
+        .with_function("fromIsoDate", |_, iso_date: LuaString| {
+            Ok(DateTime::from_iso_date(iso_date.to_string_lossy()))
+        })?
+        .with_function("toIsoDate", |_, this| Ok(DateTime::to_iso_date(&this).map_err(|()| LuaError::external(
+            "failed to parse DateTime object, invalid",
+        ))))?
+        .with_function(
+            "formatTime",
+            |_, (this, timezone, fmt_str, locale): (DateTime, LuaValue, LuaString, LuaString)| {
+                Ok(DateTime::format_time(
+                    &this,
+                    Timezone::from_lua(timezone, lua)?,
+                    fmt_str.to_string_lossy(),
+                    locale.to_string_lossy(),
+                ).map_err(|()| LuaError::external(
+                    "failed to parse DateTime object, invalid",
+                )))
+            },
+        )?
+        .build_readonly()
+}
+
+impl<'lua> FromLua<'lua> for TimestampType {
+    fn from_lua(value: LuaValue<'lua>, _: &'lua Lua) -> LuaResult<Self> {
+        match value {
+            LuaValue::Integer(_) => Ok(TimestampType::Seconds),
+            LuaValue::Number(num) => Ok(if num.fract() == 0.0 {
+                TimestampType::Seconds
+            } else {
+                TimestampType::Millis
+            }),
+            _ => Err(LuaError::external(
+                "Invalid enum type, number or integer expected",
+            )),
+        }
+    }
+}
+
+impl LuaUserData for DateTime {
+    fn add_fields<'lua, F: LuaUserDataFields<'lua, Self>>(fields: &mut F) {
+        fields.add_field_method_get("unixTimestamp", |_, this| Ok(this.unix_timestamp));
+        fields.add_field_method_get("unixTimestampMillis", |_, this| {
+            Ok(this.unix_timestamp_millis)
+        });
+
+        fields.add_field_method_set("unixTimestamp", |_, this, val| {
+            this.unix_timestamp = val;
+
+            Ok(())
+        });
+
+        fields.add_field_method_set("unixTimestampMillis", |_, this, val| {
+            this.unix_timestamp_millis = val;
+
+            Ok(())
+        });
```

Comment:
I don't think these should have lua-accessible setters, this invalidates the contract that a given DateTime is always a specific point in time, and it would also be possible to set `unixTimestamp` without setting `unixTimestampMillis` or vice-versa when they are really mutually dependent, potentially leading to unexpected behavior

```suggestion
```